client/components/migration/migration-runner.js

```diff
@@ -0,0 +1,251 @@
+    },
+    stateDeactivating: {
+        success: 'stateDone',
+        fail: 'stateErrorDeactivating',
+        callback: fetchAPICall(deactivateWCSTPluginAPICall),
+    },
+    stateErrorDeactivating: {
+        success: 'stateDeactivating',
+        fail: 'stateErrorDeactivating',
+        callback: stateErrorHandlingAPICall(MIGRATION_STATE_ENUM.ERROR_DEACTIVATING),
+    },
+    stateDone: { // Done state.
+        success: null,
+        fail: null,
+        callback: null
+    }
+}
+
+// Any of the state that halts the machine. Including "done" or any errors.
+const stopStates = ['stateDone', 'stateErrorInstalling', 'stateErrorActivating', 'stateErrorDB', 'stateErrorDeactivating'];
+
+/**
+ * This function runs the entire migration based on what the previousMigrationState is. It will stop
+ * when it lands on an "error state" or when the migration is completed.
+ *
+ * @param {number} previousMigrationState This is the "wcshipping_migration_state" stored in the option table. This is the last state recorded for this migration before it stopped.
+ */
+const installAndActivatePlugins = async(previousMigrationState) => {
+    const runNext = async (migrationState) => {
+        if (!migrationState) {
+            // Nothing to run, do nothing and return.
+            return;
+        }
+
+        const currentMigrationState = migrationStateTransitions[migrationState];
+        let nextMigrationStateToRun = '';
+        try {
+            await currentMigrationState.callback();
+            nextMigrationStateToRun = currentMigrationState.success;
+        } catch (e) {
+            nextMigrationStateToRun = currentMigrationState.fail;
+        } finally {
+            return nextMigrationStateToRun;
+        }
+    };
+
+    /**
+     * This function checks what the next state to run is. If there is no record of any previous migration run,
+     * then we will start from the beginning.
+     * If there is a record about where we last stopped at, then we start from that state's next state.
+     *
+     * @returns {string} The next state to run. The name of the state is the key in this object migrationStateTransitions.
+     */
+    const getNextStateToRun = () => {
+        if ( ! previousMigrationState) {
+            // stateInit
+            return MIGRATION_ENUM_TO_STATE_NAME_MAP[2];
+        }
+
+        const currentStateName = MIGRATION_ENUM_TO_STATE_NAME_MAP[previousMigrationState];
+        return migrationStateTransitions[currentStateName].success; // The next state is "success".
+    };
+
+    // Run the migration chain from where it last stopped. If there is no record, start from the beginning.
+    let nextMigrationStateToRun = getNextStateToRun();
+    let runAttemps = 0;
+    const maxAttempts = Object.keys(migrationStateTransitions).length; // The states don't loop, thus it can't be more than its size. Serve as a safe guard in case of infinite loop.
```

Comment:
```suggestion
    const maxAttempts = Object.keys( migrationStateTransitions ).length; // The states don't loop, thus it can't be more than its size. Serve as a safe guard in case of infinite loop.
```